python/pyspark/sql/pandas/group_ops.py

```diff
@@ -573,7 +579,11 @@ def transformWithStateUDF(
                 statefulProcessorApiClient.set_handle_state(StatefulProcessorHandleState.CLOSED)
                 return iter([])
 
-            result = handle_data_with_timers(statefulProcessorApiClient, key, inputRows)
+            batch_timestamp, watermark_timestamp = get_timestamps(statefulProcessorApiClient)
```

Comment:
Ideally this shouldn't be called at every key. If we split out the handling of timer expiration from the handling of input rows, we would only need to call this at once.

Reply:
Moved this `get_timestamps()` into `stateful_processor_api_client.py` inside `__init__()` so we will only make an API call only once for each batch.